diagnostics-app/src/app/app.component.spec.ts

```diff
@@ -6,21 +6,61 @@
  * @fileoverview Test file for the app.component.ts
  */
 
-import { TestBed } from '@angular/core/testing';
+import { ComponentFixture, TestBed } from '@angular/core/testing';
 import { RouterTestingModule } from '@angular/router/testing';
 import { AppComponent } from './app.component';
 
-describe('AppComponent', () => {
+describe('component: app', () => {
+  let fixture: ComponentFixture<AppComponent>;
+  let component: AppComponent;
+
   beforeEach(async () => {
     await TestBed.configureTestingModule({
       imports: [RouterTestingModule],
       declarations: [AppComponent],
     }).compileComponents();
   });
 
-  it('should create the app', () => {
-    const fixture = TestBed.createComponent(AppComponent);
-    const app = fixture.componentInstance;
-    expect(app).toBeTruthy();
+  beforeEach(() => {
+    fixture = TestBed.createComponent(AppComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create app component', () => {
+    // component is created
+    expect(component).toBeTruthy();
+  });
+
+  it('should set dark mode by default', () => {
+    // default theme is dark
+    expect(component.theme).toBe('dark');
+  });
+
+  it('should toggle the theme', () => {
+    // theme = 'dark' by default
+    component.toggleTheme();
+    fixture.detectChanges();
+    // css class 'light' should be added
+    expect(component.theme).toBe('light');
+    expect(component.cssClass).toBe('light');
+    // switch again
+    component.toggleTheme();
+    fixture.detectChanges();
+    // css class 'light' should be added
+    expect(component.theme).toBe('dark');
+    expect(component.cssClass).toBe('dark');
+  });
+
+  it('should toggle theme on button click', () => {
+    // simulate component.toggleTheme()
+    spyOn(component, 'toggleTheme');    
+    const buttonRef: HTMLButtonElement =
+      fixture.nativeElement.querySelector('.theme-toggle');
+    // simulate toggle theme button click
+    buttonRef.click();
+    fixture.detectChanges();
+    // should invoke the toggleTheme() function
+    expect(component.toggleTheme).toHaveBeenCalled();
```

Comment:
Let's also check the theme value was changed:
```
    // css class 'light' should be added
    expect(component.theme).toBe('light');
    expect(component.cssClass).toBe('light');
```

Reply:
I had this doubt while writing the test. Since I have already tested the toggleTheme function, wouldn't this become redundant?

Reply:
Correct.

The previous test is a unit_test, while this one is integration/component test (that's clicking on the button triggers the appropriate logic, and changes the system state as expected). Generally, we need both. Maybe we can split the two types of tests in different files? I am not familiar with Angular, what is the best practice here?

Reply:
Integration tests in terms of Angular are where two or more components are interacting together. In normal contexts this would have been an integration test. But since the button is part of the same component, it is considered to be a unit test.

If this button was a different component, it will be called an integration test.

Reply:
> In Angular, the difference between unit and integration tests is sometimes subtle. A unit test typically focusses on a single Angular Component, Directive, Service, Pipe, etc. Dependencies are replaced with fakes. An integration test spans one Component together with its children and possibly connected Services as well. It is also possible to write a test that integrates all parts of an Angular Module.

From, https://testing-angular.com/testing-principles/#distribution-of-testing-efforts

Reply:
This is a relevant example of how we can structure the tests: [Example](https://github.com/9elements/angular-workshop/blob/main/src/app/components/service-counter/service-counter.component.spec.ts)

Contrary to this, when I write integration tests, we can keep them in separate files such as:
`app.component.unit.spec.ts` and `app.component.int.spec.ts`

Reply:
Thanks for sharing. I am convinced that both are unit tests but since there are redundancy, in my opinion, we should keep the second test and remove the first one. This entails checking theme and cssClass fields after clicking the button.

Reply:
Sure, will do that.